File: app/src/main/java/com/example/tmdbpeople/viewmodels/SearchPersonsViewModel.java
package com.example.tmdbpeople.viewmodels;

import android.util.Log;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MediatorLiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;
import androidx.paging.LivePagedListBuilder;
import androidx.paging.PageKeyedDataSource;
import androidx.paging.PagedList;

import com.example.tmdbpeople.datasource.PersonDataSource;
import com.example.tmdbpeople.datasource.PersonDataSourceFactory;
import com.example.tmdbpeople.datasource.searchdatasource.PersonSearchDataSource;
import com.example.tmdbpeople.datasource.searchdatasource.PersonSearchDataSourceFactory;
import com.example.tmdbpeople.models.PersonModel;
import com.example.tmdbpeople.networkutils.Constants;

public class SearchPersonsViewModel extends ViewModel {

    private LiveData<PagedList<PersonModel>> personPagedList = new MutableLiveData<>();
    private LiveData<PageKeyedDataSource<Integer, PersonModel>> liveDataSource = new MutableLiveData<>();
    PersonSearchDataSourceFactory personDataSource;


    public SearchPersonsViewModel() {
        personDataSource = new PersonSearchDataSourceFactory(null);

        liveDataSource = personDataSource.getItemLiveDataSource();

        PagedList.Config pagedListConfig =
                (new PagedList.Config.Builder())
                        .setPageSize(PersonDataSource.PAGE_SIZE)
                        .setEnablePlaceholders(false).build();

        personPagedList = (new LivePagedListBuilder(personDataSource, pagedListConfig))
                .build();
    }

    public void doSearch(String query) {
        if (!query.isEmpty()) {
            personDataSource.setQuery(query);
            personDataSource.invalidate();
        }
    }

    public LiveData<PagedList<PersonModel>> getPersonPagedList() {
        return personPagedList;
    }
}

File: app/src/main/java/com/example/tmdbpeople/viewmodels/SearchPersonsViewModel.kt
package com.example.tmdbpeople.viewmodels

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import androidx.paging.LivePagedListBuilder
import androidx.paging.PageKeyedDataSource
import androidx.paging.PagedList

import com.example.tmdbpeople.datasource.searchdatasource.PersonSearchDataSourceFactory
import com.example.tmdbpeople.models.responsemodels.PersonDetailsResponse
import com.example.tmdbpeople.networkutils.Constants
import com.example.tmdbpeople.networkutils.LoadCallback

class SearchPersonsViewModel(var loadCallback: LoadCallback) : ViewModel() {

    var personPagedList: LiveData<PagedList<PersonDetailsResponse?>> = MutableLiveData()
    private var liveDataSource = MutableLiveData<PageKeyedDataSource<Int?, PersonDetailsResponse?>>()
    internal var personDataSource: PersonSearchDataSourceFactory

    init {
        personDataSource = PersonSearchDataSourceFactory(loadCallback,null)

        liveDataSource = personDataSource.itemLiveDataSource

        val pagedListConfig = PagedList.Config.Builder()
            .setPageSize(Constants.PAGE_SIZE)
            .setEnablePlaceholders(false).build()

        personPagedList = LivePagedListBuilder(personDataSource, pagedListConfig)
            .build()
    }

    fun doSearch(query: String) {
        if (!query.isEmpty()) {
            personDataSource.query = query
            personDataSource.invalidate()
        }
    }
}
